### PR TITLE
Rotate with 2 vectors

### PR DESCRIPTION
**Pull request by carterbox**
Implements the two vector rotation specification that I've been talking about.

Two things that need to be addressed:

1. Where do .bib references go? I included one in my documentation.
2. Tests for this new method are applied directly instead of through the `Polytope` class. Is that acceptable?

### CODE REVIEW

**Comment by carterbox** [comment]
This is failing because I used the `@` operator for [matrix multiplication](https://docs.scipy.org/doc/numpy/reference/generated/numpy.matmul.html#numpy.matmul). I don't know what version of `numpy` this was introduced.

**Comment by johnyf** [comment]
1. For reference management please see #43.

2. I think that unit tests at a granularity finer than the code structure (e.g., a class, in this case `Polytope`) are fine. It is good to write small tests, because it encourages the developer to answer the question "what is this function/method/data structure doing?". (A formal specification is one way of answering this question.)

**Comment by johnyf** [comment]
Thanks for pointing to PEP 465. Python's new operator for matrix multiplication will make numerical code more readable, including `polytope` and `tulip`. It provides motivation for dropping Python 2, but I suggest that we wait until Python 2 [EOL](https://pythonclock.org/) [PEP 373](http://legacy.python.org/dev/peps/pep-0373/). Perhaps it has also been a difference that discouraged users from adopting Python.

Presently, [`install_requires`](https://github.com/tulip-control/polytope/blob/a037b555758ed9ee736fa7cb324d300b8d622fb4/setup.py#L108) `numpy >= 1.6`. `numpy == 1.10.0` was [released on 5 Oct 2015](https://github.com/numpy/numpy/releases/tag/v1.10.0). I don't think we are significantly constraining users if we require `numpy >= 1.10.0`.

However, code that contains the operator `@` will be parsed correctly by `CPython >= 3.5` and [`tauthon`](https://github.com/naftaliharris/tauthon#matrix-multiplication-operator), not by `CPython < 3.5` (`@` [appears](http://stackoverflow.com/questions/35333234/can-one-import-operator-with-future-in-python3-4) not to be a candidate for backporting).

Based on the previous arguments, I suggest replacing `@` with a call to the function `numpy.dot`.

**Comment by carterbox** [comment]
I'm slightly annoyed with continued Python 2 support, but I imagine it has something to do with extensive libraries needing to be converted and people not having the man hours to do it.

**Comment by johnyf** [comment]
There are many systems in operation, and other priorities may be more important than a transition to Python 3. Those of the popular Python packages that have transitioned to Python 3 are tracked at http://py3readiness.org. `tulip` is in the process of transitioning to Python 3 https://github.com/tulip-control/tulip-control/issues/44.

**Comment by carterbox** [comment]
I have no more commits for this branch, just in case you were waiting for me to fix something.

**Comment by slivingston** [comment]
Regarding the second question in [the opening post](https://github.com/tulip-control/polytope/pull/42#issue-211995212), I think that it is best for unit tests about the new function to apply it directly, not via `Polytope`. After all, the function is independent of the class `Polytope`.

File: polytope/polytope.py [hunk 1]
```diff
@@ -570,6 +564,18 @@ def _rotate(polyreg, i=None, j=None, u=None, v=None, theta=None, R=None):
     return R
 
 
+def _Givens_rotation_matrix(i, j, theta, N):
```
Comment (slivingston) [inline polytope/polytope.py]: Why not make this part of the public API (and delete the leading `_`)?

Whether or not it is part of the public API, there should be unit tests created for it.

Comment (johnyf) [inline polytope/polytope.py]: Minor note about capitalization and function names (I do not find the name `givens_rotation_matrix` very readable, but for the sake of uniform naming).

Comment (carterbox) [inline polytope/polytope.py]: If all these rotation functions are added to a public API, shouldn't they be in their own module instead of with the `Polytope` class definitions?

Givens is the name of the person who introduced this rotation matrix, and as far as I know, people refer to it as the Givens rotation matrix. That's why it is named the way it is.

Comment (johnyf) [inline polytope/polytope.py]: I am aware that "Givens" is a person's name, but by PEP8 function names need to be lowercase.

Most of the module `polytope.polytope` is algorithms implemented as functions, so the contents are along the same lines. Whether it is better or not to make these functions public I am not sure. Maybe we can wait to see how frequently users need to call them directly.

In general, one problem that will arise from splitting the `polytope` module is cyclic imports. This may not be an issue for rotation functions, but it is for other functions. What would be the motivation for placing the functions in another file?

Comment (carterbox) [inline polytope/polytope.py]: Moving some of the functions to a different file would save me from having to scroll so much!

Comment (johnyf) [inline polytope/polytope.py]: Oh, but that could be addressed by opening multiple views of the same file, positioned at different locations. Using [Atom](https://atom.io), these views could be different tabs side by side, or even different tabs in the same pane. Also, the [`minimap` package](https://atom.io/packages/minimap) together with packages that show changed lines using `git` are quite convenient for accelerating navigation. For long text files (like `polytope`), editing a file while open in 2 or 3 tabs can exhibit slight lag, but that is a minor issue, and can be mitigated by switching to a different tab.

Comment (slivingston) [inline polytope/polytope.py]: I agree with splitting the file into topic-oriented files. It should be possible to do this and avoid troubles with cycle dependencies. However, I think that it is outside the scope of this PR.

**Comment by slivingston** [comment]
@carterbox can you show me example that motivated your following comment in `solve_rotation_ap`?

> the sign of arctan2 in numpy seems to be the opposite of arctan2 in \citea{aguilera2004general}.

**Comment by slivingston** [comment]
Let function arctan2 as defined by Aguilera and Pérez on page 1 of their paper be called `arctan2_ap`. I found that it is different from `numpy.arctan2` in the following respects. For (x,y) \in R^2,

* if x = y = 0, then arctan2_ap(0,0) = pi/2 whereas `numpy.arctan2(0,0) = 0.0`;
* if x < 0 and y < 0, then arctan2_ap(y, x) > 0 whereas `numpy.arctan2(y, x) < 0`

Otherwise they agree on finite values of x and y.

**Comment by carterbox** [comment]
@slivingston Do you still need me to explain my comment about `arctan2_ap`? It seems that you figured it out.

**Comment by slivingston** [comment]
no need to explain. I will finish this today.

**Comment by slivingston** [comment]
I have several final comments, which I give below. Because they concern documentation and are not critical, and because other work is depending on this PR to close, I think that we can address them after merging.

**Comment by slivingston** [comment]
Though they take different values in one quadrant as [I observe above](https://github.com/tulip-control/polytope/pull/42#issuecomment-294561281), for points away from the origin the two arctan2 functions agree when viewed as functions to S^1 (1-sphere). Because angles from arctan2 are only used in `solve_rotation_ap` to construct Givens rotation matrices, in the scope of `solve_rotation_ap` we can treat the two definitions of arctan2 as equivalent.

**Comment by slivingston** [comment]
The crucial distinction between the paper and `solve_rotation_ap` seems to be the side on which matrix multiplication is intended for the rotation matrices. In the paper and the algorithm (and in your implementation of it), rotation matrices are left-multiplied by vectors, e.g., as in

    uv = uv.dot(Mk)

whereas the rotation matrix R obtained from calling `solve_rotation_ap` should be applied by multiplying it with the vector to be rotated on the right, e.g., as in the unit tests that you created.

This distinction explains why -arctan2 is used in your implementation, rather than arctan2 (positive) as in the paper. In particular, differences in the definitions of arctan2 do not require this sign change. As such, I think that the comment in the docstring
> Also, the sign of arctan2 in numpy seems to be the opposite of arctan2 in \citea{aguilera2004general}.

should be deleted.

**Comment by slivingston** [comment]
The comment
> NOTE: This method is not implemented at this time.

in the docstring of `_rotate` can be deleted because it refers to the method now implemented in `solve_rotation_ap`.

**Comment by slivingston** [comment]
Why is the angle of rotation twice that of the angle between the vectors?

I did not ask earlier because I guessed that it would be explained in the paper or would be a consequence of the algorithm, but I found no evidence of this. E.g., `solve_rotation_ap` can be modified to have
```python
theta = -np.arctan2(uv[1, 1], uv[1, 0])
```
instead of `theta = -2 * np.arctan2(uv[1, 1], uv[1, 0])`.

**Comment by carterbox** [comment]
@slivingston The applied rotation is twice the angle between the two vectors to prevent ambiguity when the applied rotation is 180 degrees. Example: In 3D, 180 degree rotations in the xy plane and the yz plane are not the same, but would have the same full angle description.

**Comment by carterbox** [comment]
Also, I'm going fix the inconsistency with matrix multiplication such that rotation matrices are on the left. I think we decided [here](https://github.com/tulip-control/polytope/issues/34#issuecomment-275895416) that vectors should be shape `(N, 1)` and thus transformations applied to the left.
